### PR TITLE
[source-hubspot] - Handle custom objects where name collides with standard objects

### LINKED ISSUE #40679: [source-hubspot] DuplicateKey error during stream refresh when custom object has same name as standard object
### Connector Name

source-hubspot

### Connector Version

4.2.7

### What step the error happened?

Configuring a new connector

### Relevant information

Create a custom object in hubspot called "owners". Try to refresh the schema.

### Relevant log output

```shell
Internal Server Error: Duplicate key io.airbyte.protocol.models.StreamDescriptor@417723ad[name=owners,namespace=<null>,additionalProperties={}] (attempted merging values io.airbyte.protocol.models.AirbyteStream@79d017cc[name=owners,jsonSchema={"$schema":"http://json-schema.org/draft-07/schema#","type":["null","object"],"properties":{"archived":{"type":["null","boolean"]},"lastName":{"type":["null","string"]},"createdAt":{"format":"date-time","type":["null","string"]},"firstName":{"type":["null","string"]},"teams":{"type":["null","array"],"items":{"type":"object","properties":{"name":{"type":["null","string"]},"id":{"type":["null","string"]},"membership":{"type":["null","string"]}}}},"id":{"type":["null","string"]},"userId":{"type":["null","integer"]},"email":{"type":["null","string"]},"updatedAt":{"format":"date-time","type":["null","string"]}}},supportedSyncModes=[full_refresh, incremental],sourceDefinedCursor=true,defaultCursorField=[updatedAt],sourceDefinedPrimaryKey=[[id]],namespace=<null>,additionalProperties={}] and io.airbyte.protocol.models.AirbyteStream@4ad67f7c[name=owners,jsonSchema={"$schema":"http://json-schema.org/draft-07/schema#","additionalProperties":true,"type":["null","object"],"properties":{"archived":{"type":["null","boolean"]},"createdAt":{"format":"date-time","type":["null","string"]},"id":{"type":["null","string"]},"properties":{"type":["null","object"],"properties":{"hs_all_accessible_team_ids":{"type":["null","string"]},"hs_user_ids_of_all_notification_unfollowers":{"type":["null","string"]},"hs_all_team_ids":{"type":["null","string"]},"hs_read_only":{"type":["null","boolean"]},"hs_object_source_detail_1":{"type":["null","string"]},"hs_object_source_detail_3":{"type":["null","string"]},"hs_object_source_detail_2":{"type":["null","string"]},"hs_object_source_label":{"type":["null","string"]},"hubspot_team_id":{"type":["null","string"]},"hubspot_owner_id":{"type":["null","string"]},"hs_unique_creation_key":{"type":["null","string"]},"hs_object_source_id":{"type":["null","string"]},"hs_user_ids_of_all_notification_followers":{"type":["null","string"]},"hs_created_by_user_id":{"type":["null","number"]},"hs_createdate":{"format":"date-time","type":["null","string"]},"hs_was_imported":{"type":["null","boolean"]},"hs_all_owner_ids":{"type":["null","string"]},"owner_email":{"type":["null","string"]},"owner_name":{"type":["null","string"]},"hs_pinned_engagement_id":{"type":["null","number"]},"hs_user_ids_of_all_owners":{"type":["null","string"]},"hs_all_assigned_business_unit_ids":{"type":["null","string"]},"hs_lastmodifieddate":{"format":"date-time","type":["null","string"]},"hubspot_owner_assigneddate":{"format":"date-time","type":["null","string"]},"hs_object_id":{"type":["null","number"]},"hs_shared_user_ids":{"type":["null","string"]},"hs_object_source":{"type":["null","string"]},"hs_merged_object_ids":{"type":["null","string"]},"hs_shared_team_ids":{"type":["null","string"]},"hs_object_source_user_id":{"type":["null","number"]},"hs_updated_by_user_id":{"type":["null","number"]}}},"updatedAt":{"format":"date-time","type":["null","string"]}}},supportedSyncModes=[full_refresh, incremental],sourceDefinedCursor=true,defaultCursorField=[updatedAt],sourceDefinedPrimaryKey=[[id]],namespace=<null>,additionalProperties={}])
```

### Contribute

- [X] Yes, I want to contribute

### PR DESCRIPTION
**Pull request by jtruty**
## What
This handles custom object names that collide with standard object names. Currently if this is the case, stream discovery fails.
Fixes: https://github.com/airbytehq/airbyte/issues/40679

I re-opened this PR from a non-org github account to allow maintainers to update the PR. This one: https://github.com/airbytehq/airbyte/pull/41639 is the original and will be closed.

## How
We check the custom object names against a set of standard hubspot object names. If there is a collision, we use the fully qualified name for the custom object instead of the actual object name.

## Review guide
airbyte-integrations/connectors/source-hubspot/source_hubspot/source.py

## User Impact
Stream discovery should work when this happens and allow a normal sync to work.

## Can this PR be safely reverted and rolled back?
- [X] YES 💚
- [ ] NO ❌

### CODE REVIEW

**Comment by jtruty** [comment]
@marcosmarxm I opened this from a non-org GH to allow maintainer edits.

File: docs/integrations/sources/hubspot.md [hunk 1]
```diff
@@ -337,6 +337,7 @@ The connector is restricted by normal HubSpot [rate limitations](https://legacyd
 
 | Version | Date       | Pull Request                                             | Subject                                                                                                                                                                          |
 |:--------|:-----------| :------------------------------------------------------- |:---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
+| 4.2.13 | 2024-07-18 | [41639](https://github.com/airbytehq/airbyte/pull/41639) | Handle custom object name collisions with standard objects |
```
Comment (marcosmarxm) [inline docs/integrations/sources/hubspot.md:340]: It is missing the metadata.yaml and pyproject connector version update.  Please fix that.

**Review by marcosmarxm — CHANGES_REQUESTED** [review]

File: docs/integrations/sources/hubspot.md (hunk 1 above)
Comment (jtruty) [reply docs/integrations/sources/hubspot.md:340]: @marcosmarxm updated!

**Comment by jtruty** [comment]
@marcosmarxm @airbytehq/dev-python any more thoughts on this?

**Comment by jtruty** [comment]
@marcosmarxm the files are updated and conflicts fixed - anything else I can do here?

File: airbyte-integrations/connectors/source-hubspot/source_hubspot/source.py [hunk 2]
```diff
@@ -220,10 +220,48 @@ def streams(self, config: Mapping[str, Any]) -> List[Stream]:
 
         return available_streams
 
+    # set of standard stream names to check for custom object name collisions
+    STANDARD_STREAM_NAMES = {
```
Comment (lazebnyi) [inline airbyte-integrations/connectors/source-hubspot/source_hubspot/source.py:224]: let's move this name to separate (e.g. utils) file and import after it there

File: airbyte-integrations/connectors/source-hubspot/source_hubspot/streams.py [hunk 3]
```diff
@@ -2245,9 +2245,10 @@ class CustomObject(CRMSearchStream, ABC):
     primary_key = "id"
     scopes = {"crm.schemas.custom.read", "crm.objects.custom.read"}
 
-    def __init__(self, entity: str, schema: Mapping[str, Any], fully_qualified_name: str, custom_properties: Mapping[str, Any], **kwargs):
+    def __init__(self, entity: str, name: str, schema: Mapping[str, Any], fully_qualified_name: str, custom_properties: Mapping[str, Any], **kwargs):
         super().__init__(**kwargs)
         self.entity = entity
+        self.name = name
```
Comment (lazebnyi) [inline airbyte-integrations/connectors/source-hubspot/source_hubspot/streams.py:2251]: we already have getter for `name` attribute, so better to proceed with this logic and add setter

File: airbyte-integrations/connectors/source-hubspot/unit_tests/test_streams.py [hunk 4]
```diff
@@ -407,6 +407,7 @@ def test_custom_object_stream_doesnt_call_hubspot_to_get_json_schema_if_availabl
 ):
     stream = CustomObject(
         entity="animals",
+        name="animals",
```
Comment (lazebnyi) [inline airbyte-integrations/connectors/source-hubspot/unit_tests/test_streams.py:410]: no need as `name` getter return `self.entity`

File: airbyte-integrations/connectors/source-hubspot/source_hubspot/source.py [hunk 5]
```diff
@@ -220,10 +220,48 @@ def streams(self, config: Mapping[str, Any]) -> List[Stream]:
...
+        "tickets",
+        "ticket_pipelines",
+        "workflows",
+    }
+
     def get_custom_object_streams(self, api: API, common_params: Mapping[str, Any]):
         for entity, fully_qualified_name, schema, custom_properties in api.get_custom_objects_metadata():
             yield CustomObject(
                 entity=entity,
+                name=fully_qualified_name if entity in self.STANDARD_STREAM_NAMES else entity,
```
Comment (lazebnyi) [inline airbyte-integrations/connectors/source-hubspot/source_hubspot/source.py:264]: could you add a comment with examples for fully_qualified_name and entity to show the difference and provide more clarity?

**Review by lazebnyi — CHANGES_REQUESTED** [review]

File: airbyte-integrations/connectors/source-hubspot/source_hubspot/source.py (hunk 5 above)
Comment (jtruty) [reply airbyte-integrations/connectors/source-hubspot/source_hubspot/source.py]: Yes, done

File: airbyte-integrations/connectors/source-hubspot/source_hubspot/streams.py (hunk 3 above)
Comment (jtruty) [reply airbyte-integrations/connectors/source-hubspot/source_hubspot/streams.py]: I modified the name getter logic instead - this is much cleaner. Let me know what you think.

**Comment by natikgadzhi** [comment]
@jtruty thank you for this! Clean PR description, and strong work. @lazebnyi will be back on Monday, we'll take a look. Looks like this is close to wrapping up.

**Comment by aldogonzalez8** [comment]
@jtruty  Change seems straightforward, but can you add a unit test for this scenario? Probably "[airbyte-integrations/connectors/source-hubspot/unit_tests/test_streams.py](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connectors/source-hubspot/unit_tests/test_streams.py)" is a good place, if you a look to  [test_custom_object_stream_doesnt_call_hubspot_to_get_json_schema_if_available](https://github.com/airbytehq/airbyte/blob/c67c5ba4e3daf4ef4b8ad764cda002d7d36c4e39/airbyte-integrations/connectors/source-hubspot/unit_tests/test_streams.py#L405-L414) should be easy to add have one for your scenario.

Also, you have a conflict.

Thanks!

**Pull request closed without merging**